micro-rdk/src/common/ota.rs

```diff
@@ -0,0 +1,359 @@
+                } else {
+                    log::info!("data length {}", data.len());
+                    #[cfg(feature = "esp32")]
+                    {
+                        let mut loader = EspFirmwareInfoLoader::new();
+                        loader.load(&data).map_err(OtaError::EspError)?;
+                        let new_fw = loader.get_info().map_err(OtaError::EspError)?;
+                        log::info!("current firmware: {:?}", running_fw_info);
+                        log::info!("new firmware: {:?}", new_fw);
+                        if let Some(ref running_fw) = running_fw_info {
+                            if running_fw.version == new_fw.version
+                                && running_fw.released == new_fw.released
+                            {
+                                log::info!("current firmware is up to date");
+                                update_handle.abort()?;
+                                return Ok(());
+                            }
+                        }
+                    }
+                    #[cfg(not(feature = "esp32"))]
+                    {
+                        if let Ok(decoded) = bincode::decode_from_slice::<
+                            EspAppDesc,
+                            bincode::config::Configuration,
+                        >(
+                            &data[..256], bincode::config::standard()
+                        ) {
+                            log::info!("{:?}", decoded.0);
+                        }
+                    }
+                    got_info = true;
+                }
+            }
+
+            if data.len() + nwritten <= OTA_MAX_IMAGE_SIZE {
+                // TODO(RSDK-9271) add async writer for ota
+                #[cfg(feature = "esp32")]
+                update_handle
+                    .write(&data)
+                    .map_err(|e| OtaError::Other(e.to_string()))?;
+                #[cfg(not(feature = "esp32"))]
+                let _n = update_handle
+                    .write(&data)
+                    .await
+                    .map_err("failed to write data to partition".to_string())?;
+                // TODO change back to 'n' after impl async writer
+                nwritten += data.len();
+            } else {
+                log::error!("file is larger than expected, aborting");
+                #[cfg(feature = "esp32")]
+                update_handle.abort()?;
+                return Err(OtaError::Other("download be weird".to_string()));
+            }
+        }
+
+        log::info!("ota download complete");
+
+        if nwritten != file_len {
+            log::error!("wrote {} bytes, expected to write {}", nwritten, file_len);
+            log::error!("aborting ota");
+            #[cfg(feature = "esp32")]
+            update_handle.abort()?;
+
+            return Err(OtaError::Other(
+                "nbytes written did not match file size".to_string(),
+            ));
+        }
+
+        drop(conn);
```

Comment:
could we drop connection before aborting the OTA? It seems to me that we'd like to keep the connection alive for as little time as possible and the abort might take a variable amount of time

Reply:
sure, I can move this up a bit to before the `nwritten != file_len` check.